### ISSUE TITLE
Allow Custom Optimizer During Training

### ISSUE THREAD

**Issue by AtomicCactus** [ultralytics/ultralytics#6260]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

In the `ultralytics/engine/trainer.py` there is a conditional that allows the user to select from several supported optimizers:

```python
if name in ('Adam', 'Adamax', 'AdamW', 'NAdam', 'RAdam'):
      optimizer = getattr(optim, name, optim.Adam)(g[2], lr=lr, betas=(momentum, 0.999), weight_decay=0.0)
  elif name == 'RMSProp':
      optimizer = optim.RMSprop(g[2], lr=lr, momentum=momentum)
  elif name == 'SGD':
      optimizer = optim.SGD(g[2], lr=lr, momentum=momentum, nesterov=True)
```

The proposal would be to overload the `optimizer=` argument of the `train()` method to either accept a `str` or an instance of `torch.optim.optimizer.Optimizer` to allow users to pass in a custom `Optimizer` instance. The trainer code could then check whether the optimizer is compatible with the scheduler and similar things and throws an error if there's an issue.

### Use case

There are newer optimizers out there, such as `Adan` and `AdaBelief`, with more coming out almost monthly. It would be really great to be able to benchmark these newer optimizers against the tried and true `AdamW` or `SGD` with YOLO models, as some of these optimizers claim faster convergence properties.

Another reason to use a custom optimizer would be to "optimize the optimizer", in other words using techniques like `torch.compile()`, `torch.jit.script()`, or `torch._foreach` or writing custom CUDA kernels either in raw form or using something like the Taichi framework to speed up the optimizer itself.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@AtomicCactus thanks for bringing this up, and I appreciate your thorough explanation of the benefits of allowing custom optimizers in YOLOv8.

The idea of allowing users to incorporate experimental optimizers for benchmarking performance against established ones is indeed valuable for the research and developer community. However, integrating external or custom optimizers into the training pipeline would require careful design to ensure compatibility with the existing framework, especially with regards to learning rate scheduling and other optimizer-dependent functionalities.

Nonetheless, we welcome contributions that could enhance flexibility while maintaining the integrity and robustness of the training process. If users or contributors are interested in experimenting with or integrating new optimizers, we would encourage them to fork the repository and work on these features within their copies.

In such cases, to maintain stability, they would need to ensure that the external optimizer conforms to the expected behavior of `torch.optim.Optimizer`. Additional code might be necessary to include checks for compatibility with learning rate schedulers and to gracefully handle any exceptions that arise from the use of untested optimizer configurations.

If you or anyone else is interested in contributing such a feature via a pull request, we recommend discussing the proposed changes in detail through an issue on GitHub first. This approach allows us to plan and review the implications of the update, as well as its potential impact on other users.

For detailed information about training customization and advanced configuration, please consult our documentation at https://docs.ultralytics.com. Your suggestions and potential contributions are valued, and they play a significant role in advancing the capabilities of YOLOv8.
